test/puppetlabs/puppetdb/query/ast_test.clj

```diff
@@ -0,0 +1,132 @@
+(ns puppetlabs.puppetdb.query.ast-test
+  "Tests the behavior of AST queries."
+  (:require
+   [clj-time.coerce :as coerce-time]
+   [clj-time.core :refer [now]]
+   [clojure.test :refer :all]
+   [puppetlabs.kitchensink.core :refer [dissoc-in]]
+   [puppetlabs.puppetdb.cli.services :as cli-svcs]
+   [puppetlabs.puppetdb.command :refer [enqueue-command]]
+   [puppetlabs.puppetdb.examples :as examples]
+   [puppetlabs.puppetdb.testutils :refer [block-until-results]]
+   [puppetlabs.puppetdb.testutils.cli :refer [get-catalogs get-factsets]]
+   [puppetlabs.puppetdb.testutils.services :as svc-utils :refer [*server*]]
+   [puppetlabs.trapperkeeper.app :refer [get-service]]))
+
+(deftest ast-queries
```

Comment:
So, http/index_test.clj sort of behaves this way - that is, most tests are generic and test cross-functionality, but has the PQL tests alongside their AST equivalent. However its full stack, involving the HTTP side of things also. So in a way, its doing a lot of the cross-entity conceptual testing already, like you have got here ... but its implementation is slightly different.

Perhaps we need to add to, improve and fix up that space rather than forking out a new test area on its own, otherwise I can just see indecision as to where tests should live becoming a thing with two generic coverage places existing. Also, the PQL coverage in the other file is already there, and this is important to ensure AST/PQL do equivalent things.

Maybe we could revise http/index_test.clj to be coverage for full-stack and direct also, who knows.

There is also room to add unit testing to certain things here, in query_eng_test.clj remember, like if you wanted to make sure certain SQL compilation/translation occurred outside of the full stack, that place is a good one. I noticed you've done some tiny fixes there, but there have been times doing unit tests there for complex query eng handling wasn't a bad idea.

Anyway, I'm not against the conceptual idea you have here, I'm just wary this is 'yet another place to add tests' and will create some ambiguity and duplication problems.

Also, query/ast_test.clj is kind of in the 'query' namespace, which we hope to remove in the future btw ... query.clj has a lot of legacy stuff in it, our goal is to collapse that functionality when we can, so probably this directory isn't go to add to atm :-).